### ISSUE TITLE
unexpected rendering triggered while yield an async request

### ISSUE THREAD

**Issue by lyhistory** [mobxjs/mobx#2715]
**Intended outcome:**
here is the codes and explanation what I've found:
```
class testStore {
  @observable tableData = []
  @observable requestComplete = false
  @action
  retrieveTableData = (flow(function * () {
    try {
      console.log('Mark 1')
      this.requestComplete = false
      console.log('Mark 2')
      const tempTableData = yield axios.post('someEndpoints','someParam')
      console.log('Mark 3')
      if (Array.isArray(tempTableData)) {
        this.tableData = doSthTo(tempTableData)
        console.log('Mark 4')  
      }
      this.requestComplete = true
      console.log('Mark 5')
    } catch (error) {
	  console.log('Mark -1')
      this.requestComplete = true
    }
  }))
}

@inject('testStore')
@observer
class testTable extends Component {
    componentDidMount () {
    const { testStore } = this.props
    testStore.retrieveTableData()
  }
  render () {
    console.log('Mark render...')
    return <Sth>
  }
```
at the very first time visiting the page testTable, output is expected:
```
Mark render...
Mark 1
Mark 2
Mark 3
Mark 4
Mark 5
Mark render...
```
but when I try to switch to other pages and coming back to this testTable again, 
I thought the output should be the same, but the actual output is quite unexpected:

**Actual outcome:**

```
Mark render...
Mark 1
Mark 2
Mark render...
Mark 3
Mark 4
Mark 5
Mark render...
```
you can see there is one more unexpected render in the middle, and I tried debug step by step, what I observed is that after the code executing sequentially to the line:
`const tempTableData = yield axios.post('someEndpoints','someParam')`
then magically it jumps back to the start:
`retrieveTableData = (flow(function * () {`
which is the @action function declaration , and after that it somehow jumps into the testTable render() , 
and after all this, it again goes back to 
`const tempTableData = yield axios.post('someEndpoints','someParam')`
and this time it just continued execute sequentially  from here till the end

**My understanding**
just this morning, an idea comes pops up, and I think I might know why it happened like that, it might be related to the Single thread Non-blocking model of the modern javascript, the main thread is single thread, but it is non-blocking, it won't just wait on this line:

`const tempTableData = yield axios.post('someEndpoints','someParam')`

it will try to do something else which is ready to do, and before this line we have:

`this.requestComplete = false` , 
btw, now this.requestComplete is true because previously the first time we visiting the page it was set to true at the end of retrieveTableData function, so now set it to false is a state update, and the thread will go ahead to do the rendering since it got nothing to do at the moment,

if my understanding is correct, I don't know is this a bug or expected behavior, from your documentation, I think strictly by right the whole action function will only trigger rendering after it's done not in the middle of the action.

**Versions**
```
"create-react-context": "^0.3.0",
"babel-preset-mobx": "^2.0.0",
"babel-preset-react-app": "^9.0.0",
"mobx": "^5.11.0",
"mobx-react": "^6.1.1",
```

**Comment by FredyC** [comment]
Can you please supply runnable reproduction?

**Comment by urugator** [comment]
More or less you got it right. You don't need `action` when using `flow`... the point of `flow` is to automatically insert `actions` **in between** time-separated blocks of code - that is - in between individual `await`s - which are replaced by `yield`s (which allows us to insert these actions).
It's not possible to batch mutations that doesn't occur synchronously. If you would need to do that (unlikely), you can use non-observable state (eg function's local vars) to track the progress and mutate the actual observable state at the end.

**Comment by mweststrate** [comment]
Or, instead of applying the changes directly, save them in a closure and
postpone the aplication: "this.name = 'hi'" -> "cont updateName = () => {
this.name = "hi" }; await sometThing(); updateName()". If there are many
changes to postpone, having an array of functions that are all executed in
the end works great.

**Comment by lyhistory** [comment]
thanks for the points, actually after some thoughts, I feel the "extra render" is absolutely correct, because we'll use requestComplete  as a flag to control some sort of "loading status", it totally make sense to update the state and trigger render while making the async call, otherwise, it's not possible to update the UI status during the async request call, I would like to close this ticket, thanks
